src/securemul.rs

```diff
@@ -0,0 +1,430 @@
+        F: FnMut(Vec<St::Item>) -> Fut,
+        Fut: Future<Output = Result<St::Item, BoxError>>,
+    {
+        type Item = St::Item;
+
+        fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            let mut this = self.as_mut().project();
+
+            loop {
+                // if future is set we poll it first before taking anything off the input stream
+                if let Some(fut) = this.future.as_mut().as_pin_mut() {
+                    let item = ready!(fut.poll(cx));
+                    this.future.set(None);
+
+                    if let Err(e) = item {
+                        // TODO (alex): we should propagate errors back to caller
+                        error!({ e }, "An error occurred computing next stream element");
+                        return Poll::Ready(None);
+                    }
+                    let item = item.unwrap();
+                    this.buf.push(item.clone());
+
+                    return Poll::Ready(Some(item));
+                } else if let Some(item) = ready!(this.stream.as_mut().poll_next(cx)) {
+                    // otherwise we poll the input stream
+                    this.buf.push(item);
+                    if this.buf.len() == *this.capacity {
+                        let items = mem::replace(this.buf, Vec::with_capacity(2));
+                        this.future.set(Some((this.f)(items)));
+                    }
+                } else if !this.buf.is_empty() {
+                    // Input stream is closed, but we still have some items to process
+                    let items = mem::take(this.buf);
+                    this.future.set(Some((this.f)(items)));
+                } else {
+                    return Poll::Ready(None);
+                }
+            }
+        }
+    }
+
+    impl<St, F, Fut> ChunkScan<St, F, Fut>
+    where
+        St: Stream,
+        F: FnMut(Vec<St::Item>) -> Fut,
+        Fut: Future<Output = Result<St::Item, BoxError>>,
+    {
+        pub fn new(stream: St, capacity: usize, f: F) -> Self {
+            Self {
+                stream,
+                capacity,
+                buf: Vec::with_capacity(capacity),
+                f,
+                future: None,
+            }
+        }
+    }
+
+    /// Consumes the input stream of replicated secret shares and produces a new stream with elements
+    /// being the product of items in the input stream. For example, if (a, b, c) are elements of the
+    /// input stream, output will contain two elements: (a*b, a*b*c)
+    ///
+    /// ## Panics
+    /// Panics if one of the internal invariants does not hold.
+    pub fn secure_multiply<'a, F, R, S>(
+        input_stream: S,
+        ctx: &'a ProtocolContext<'a, R>,
+        index: u128,
+    ) -> impl Stream<Item = ReplicatedSecretSharing<F>> + 'a
+    where
+        S: Stream<Item = ReplicatedSecretSharing<F>> + 'a,
+        F: Field + 'static,
+        R: Ring,
+    {
```

Comment:
I don't really understand what is going on here, and how (if at all), I as a protocol author need to interact with Streams.